ユーザーフォロー機能の課題を提出させていただきます。

kotokotodonabe opened the pull request.

Commit:
feat: ユーザーフォロー機能の実装

add:
中間テーブルFollowRelationshipテーブルの作成
フォロー機能、フォロー解除機能
フォロー、フォロワー一覧

Commit:
fix: follow_relationshipsテーブルで外部キーをuniqueにしており、同じユーザーをフォローできなかったためテーブルを修正した。

修正した手順
1. drop_table_follow_relationshipという名前のmigrationファイルを作成し、drop_table :follow_relationshipsとした。
2. rails db:migrateを実行。
3. follow_rerationshipのmigrationファイルを修正した状態で再度作成し、migrationファイルを実行。

Commit:
fix: pry-railsをdevelopment環境ではないところでインストールしていたので削除

Commit:
自分自身の事もフォローできてしまっていたため、フォローできないようにした。

Commit:
不要なファイルの削除

Commit:
rubocopとlintを通した。

betachelsea commented on app/controllers/follow_relationships_controller.rb:
```diff
@@ -0,0 +1,15 @@
+  # フォローするとき
+  def create
+    current_user.follow(params[:user_id])
+    redirect_to request.referer
```
[nits] もしAPIが公開されている状態(ログインなどなしで、誰でも好きにリクエストできる)の場合は、想定外のところからリクエストが飛んできます。そういう場合にrefererを利用すると、どんな値が入ってくるかわからないので要注意です。今回はログイン必須なので事実上は問題ないかもしれませんが、安易にrefererが使われている様子をみると、あとから何か脆弱性が発生していないか気になってしまいます。よければ別の方法に差し替えてもらいたいです〜 🙏

betachelsea commented on db/migrate/20221027001757_drop_table_follow_relationship.rb:
```diff
@@ -0,0 +1,5 @@
+class DropTableFollowRelationship < ActiveRecord::Migration[6.1]
+  def change
+    drop_table :follow_relationships
```
[MUST] `bin/rails db:migration` を実行したところ、このmigrationファイルの実行時に失敗してしまいました。 :eyes:  もうひとつの `db/migrate/20221027002048_create_follow_relationships.rb` の前に、一度 `follow_relationships` テーブルを作る → 消す → 再作成 でやりなおしがあったのかな？と思いました。

リポジトリの中のmigrationファイルは、mainブランチ上で一貫して実行しても問題ないように維持するのが肝要です。そうしておくことで、途中でjoinしたチームメンバーが環境構築するときにスムーズにできますし、実際にサービスでトラブルがあった時にDB再構築してデータを引っ越したり(滅多にやりたくないものですが)する時に困らずにすみます。

もし、この `db/migrate/20221027001757_drop_table_follow_relationship.rb` ファイルが一時的な対応のためだけのもので実際は不要とのことでしたら、削除しておいてもらいたいです〜 🙏

betachelsea commented on app/controllers/users_controller.rb:
```diff
@@ -7,5 +7,19 @@ def index
+  # フォローワー一覧
+  def follows
+    user = User.find(params[:id])
+    @users = user.followings.page(params[:page]).per(3)
```
[nits] `.per(3)` は、実際には1ページ3人ずつの表示だと少ないと思うので、もっと多めにしていくのがいいですね。また、`3` はマジックナンバーなので、定数にする or kaminari の `paginates_per` の機能でモデル側に定義するのも一案ですね。


[kaminari/kaminari: ⚡ A Scope & Engine based, clean, powerful, customizable and sophisticated paginator for Ruby webapps](https://github.com/kaminari/kaminari#configuring-default-per_page-value-for-each-model-by-paginates_per)

> Configuring Max per_page Value for Each Model by max_paginates_per

betachelsea commented on app/models/user.rb:
```diff
@@ -5,4 +5,31 @@ class User < ApplicationRecord
+  has_many :followings, through: :follower, source: :followed
+
+  # ユーザーをフォローした時の処理
+  def follow(user_id)
```
[nits] user_id (数字)を直接もらう形だと、そのidの User が本当にあるのかどうかがわからないので、もらった後か前かでUserのインスタンスを確認できる実装になっているといいですね。

betachelsea replied:
[IMO] 今回、user_idが渡される前にインスタンス確認をしてもらう形にしていただきましたね！
一点問題として、それだとこの `follow` 利用前に意識的にインスタンスを確認する、というフローを必ずいれないとという縛りを意識的に考えておかないといけなさそう、というのがちょっと心配かなと思いました。

対応案として、次のようなやり方があるよというのを共有しておきます〜。

```ruby
# もらう時はインスタンスをもらうようにする (すると、与える時に User.find とかで確認するようになる)
def follow(user)
   follower.find_or_create_by(followed_id: user.id)
   # もしくは follower.find_or_create_by(followed: user)  でもいけると思います
end
```

betachelsea commented on app/models/user.rb:
```diff
@@ -5,4 +5,31 @@ class User < ApplicationRecord
+
+  # フォローしているか判定
+  def following?(user)
+    followings.include?(user)
```
[nits] `include?` だと全件走査になりそうなので、 `exists?` がオススメです〜。
参考: [ActiveRecord::FinderMethods - exists?(conditions = :none)](https://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-exists-3F)

betachelsea commented on app/views/users/followers.html.erb:
```diff
@@ -0,0 +1,13 @@
+<h1>フォロワー一覧</h1>
+
+<% if @users.exists? %>
+  <% @users.each do |user| %>
```
[nits] `app/views/users/follows.html.erb` でも同じように定義されているので、partial化して共通に使用しても良さそうですね！

betachelsea commented on app/models/follow_relationship.rb:
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class FollowRelationship < ApplicationRecord
```
[MUST] 現在、次の手順で操作すると、Aさんが何回もBさんをフォローする、ということができるようです。

- Aさんとしてログイン
- Bさんのページを複数タブ(1, 2とする)で開く
- 1のタブでBさんをフォローする
- 2のタブでBさんをフォローする (できてしまう)
- フォロー・フォロワーのページの状況をみると、複数回名前が載ってしまう

validationを使って、フォロー・フォロワー関係の組み合わせが全く同じデータが作成されないようにしてみましょう。

betachelsea commented on app/models/user.rb:
```diff
@@ -5,4 +5,31 @@ class User < ApplicationRecord
+
+  # ユーザーをフォローした時の処理
+  def follow(user_id)
+    follower.create(followed_id: user_id)
```
[MUST] 2タブで2回フォローする、という操作をすると、validationが正常にできても、別の意味で例外エラーとなってしまいそうです。例外エラーは出ないようにしてみていただきたいです 🙏 
ヒント:RailsのActiveRecordでは「データが見つからなければ作る」といった操作をするメソッドもありますので、探してみると良さそうです。

betachelsea commented on app/models/user.rb:
```diff
@@ -5,4 +5,31 @@ class User < ApplicationRecord
+
+  # ユーザーのフォローを外した時の処理
+  def unfollow(user_id)
+    follower.find_by(followed_id: user_id).destroy
```
[MUST] こちらも、2タブで2回フォローを外す、という操作で例外エラーとなります。例外エラーは出ないようにしていただきたいです 🙏 
ヒント: 「safe navigation」を調べてみましょう。

betachelsea commented on app/views/users/show.html.erb:
```diff
@@ -1,4 +1,12 @@
+  <% if current_user.following?(@user) %>
+    <%= link_to 'フォロー外す', user_follow_relationships_path(@user.id), method: :delete, class: "btn-sm btn-danger" %>
+  <% else %>
+    <%= link_to 'フォローする', user_follow_relationships_path(@user.id), method: :POST, class:"btn-sm btn-success" %>
```
[FYI] method表記は `:post` と小文字でも大丈夫です。 ( `:delete` が小文字だったので、揃えないのかな〜と思って気になりました :eyes: )

betachelsea commented on db/migrate/20221027002048_create_follow_relationships.rb:
```diff
@@ -0,0 +1,12 @@
+class CreateFollowRelationships < ActiveRecord::Migration[6.1]
+  def change
+    create_table :follow_relationships do |t|
+      t.integer :follower_id
+      t.integer :followed_id
```
[MUST] nullが入ると困るので、nullを許さないようにするのがいいですね！　
[nits] また、"外部キー設定"と"ユニーク制約"を入れておくのも良さそうです。

betachelsea reviewed (commented):
提出ありがとうございます！コメントしました〜 📝

kotokotodonabe commented:
@betachelsea 
レビューいただきありがとうございます！

Commit:
フォロー、アンフォローした際のリダイレクト先を変更した。

redirect.refererをリダイレクト先にしていたが脆弱性があるので、フォロー、アンフォローした対象のユーザーのインスタンス変数を用意して、そこにリダイレクトするようにした。

Commit:
paginates_perを使用してデフォルトで表示する件数を設定した。

マジックナンバーを使用してユーザーを表示してしまっていたが、モデル側で表示するデフォルトの件数を設定した。

Commit:
全表スキャンになっていたため、それを修正した。

exists?メソッドを使用することにより、条件に該当するレコードが存在した瞬間に探索を終了してくれる。

Commit:
フォロー、フォロワー一覧をパーシャルファイルに分けた。

Commit:
タブを複数開いてフォロー機能を使用した際のバグの修正

タブを二つ開いた時に同じユーザーをフォローできてしまう挙動を修正。
フォロー、アンフォロー機能を使用した際の例外エラーを出力させないように修正。

Commit:
メソッド名を小文字にした。

Commit:
follow_relationshipsのカラムのオプションを追加した。

Commit:
ユーザーをフォローした際の挙動を変更した。

ユーザー詳細ページを開いている時に、対象のユーザーがアカウントを削除した場合、フォローせずにユーザー一覧ページへリダイレクト

betachelsea commented on app/controllers/follow_relationships_controller.rb:
```diff
@@ -0,0 +1,27 @@
+      current_user.follow(params[:user_id])
+      redirect_to @other_user
+    else
+      flash[:alert] = "このユーザーは存在しません。"
```
とても丁寧な対応ですね！ 👍 
別案ですが、ここでユーザーが存在しない = 不正なリクエストと判断して、丁寧にやらずにエラーで返してよし、というやり方もあります！共有まで〜 🙏 

例: `User.find(params[:user_id])` で探して、存在しなければ例外( ActiveRecord::RecordNotFound ) が出る。ここで例外が出るというのはある意味正しいとして、特にエラーメッセージなどは対応しない。

メンターのjnchitoさんのこの記事は、とてもわかりやすくてオススメです！
[Railsアプリケーションにおけるエラー処理（例外設計）の考え方 \- Qiita](https://qiita.com/jnchito/items/3ef95ea144ed15df3637)

kotokotodonabe replied:
@betachelsea 
ありがとうございます！！

betachelsea commented on app/controllers/follow_relationships_controller.rb:
```diff
@@ -0,0 +1,27 @@
+  def create
+    if @other_user = User.exists?(id: params[:user_id])
+      current_user.follow(params[:user_id])
+      redirect_to @other_user
```
[MUST] ううむ、ここって動きましたか？  `@other_user` には、 `User.exists?` の結果の True/False が入っており、Userのインスタンスがないので、どこにもうまくリダイレクトされなさそうです。再度確認をお願いします！ 🙏

betachelsea commented on app/models/follow_relationship.rb:
```diff
@@ -0,0 +1,9 @@
+
+class FollowRelationship < ApplicationRecord
+  validates :follower, uniqueness: { scope: [:follower_id, :followed_id] }
+  validates :followed, uniqueness: { scope: [:follower_id, :followed_id] }
```
[FYI] 2個validatesが定義されていると、エラーになった時に2つ警告が出る点には要注意かもしれません〜 :eyes: 

```ruby
# すでに follower_id: 1、 followed_id: 2 のデータがあるとして
item = FollowRelationship.new(follower_id: 1, followed_id: 2)
item.valid?  
#=> false

item.errors.full_messages
#=> ["Followerはすでに存在します", "Followedはすでに存在します"]
```

主にエラーは能動的に実行した側の視点で出るのが良いかなと思うので、

```ruby
validates :followed, uniqueness: { scope: :follower_id }
```

とするとスッキリするかなと思いました〜 👀   (エラーは `["Followedはすでに存在します"]` のみとなる)

※ Followed を フォロー対象者と解釈してましたが、逆だったらすみません 🙇

betachelsea commented on app/views/users/_follow_relationship.html.erb:
```diff
@@ -0,0 +1,13 @@
+<h1><%= "#{title}" %>一覧</h1>
```
[nits] `<h1>...</h1>` は 、
- `app/views/users/followers.html.erb`
- `app/views/users/follows.html.erb` 

にそれぞれ残置しておくとわかりやすいかと思います！

betachelsea commented on app/views/users/_follow_relationship.html.erb:
```diff
@@ -0,0 +1,13 @@
+<h1><%= "#{title}" %>一覧</h1>
+
+<% if @users.exists? %>
```
[nits] パーシャル内でインスタンス変数 ( `@users` など )を使うと、該当パーシャルを呼び出す時に必ず定義しなければいけない、という制限ができてしまう(依存性が高まる)ので、 インスタンス変数はさけて ローカル変数を渡すようにすることをオススメします！ `title` は渡せていますね。 👀

betachelsea commented on db/migrate/20221027002048_create_follow_relationships.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :follow_relationships do |t|
+      t.integer :follower_id, null: false, foregin_key: true
+      t.integer :followed_id, null: false, foregin_key: true
```
[nits] `foreign_key` かな？と思いました 👀  また、`foreign_key` オプションは reference型 でないと使用できないようです 🙏 

[Railsの外部キー制約とreference型について \- Qiita](https://qiita.com/ryouzi/items/2682e7e8a86fd2b1ae47)

betachelsea commented on app/views/users/show.html.erb:
```diff
@@ -33,6 +41,15 @@
+    <h6>フォロー<%= @followings.count %></h6>
+  <% end %>
+  <%= link_to followers_user_path(@user) do %>
+    <h6>フォロワー<%= @followers.count %></h6>
```
[nits] 改めて見てちょっと気になったので… :eyes: 
`<h5>` など他の見出しレベルから飛ばして、急に `<h6>` を使うのは不自然ですね。もしフォントサイズなどを調整したい場合は、CSSで対応していただきたいです 🙏 

参考: [アウトラインとは？ \| FBC  https://bootcamp.fjord.jp/pages/420](https://bootcamp.fjord.jp/pages/420)

betachelsea commented on app/views/users/_follow_relationship.html.erb:
```diff
@@ -0,0 +1,13 @@
+  <% @users.each do |user| %>
+    <div class="one-post card w-100">
+      <%= link_to user_path(user) do %>
+        <p class="caption h4"><%= user.name %></p>
```
[nits] `<p>` を使っていますが、文書では無さそうなので、 `caption h4` のclassをあてたいということであれば、 aタグ( link_to )にあてるのがよさそうです〜。

```ruby
<%= link_to user_path(user), class: "caption h4" do %>
  <%= user.name %>
<% end %>

# これだったら↓にしたほうがスッキリですね！

<%= link_to user.name, user_path(user), class: "caption h4" %>
```

横並びになってお困りのようであれば、一覧なので、`<ul>` 使うのもアリかもしれません〜。

betachelsea reviewed (commented):
修正ありがとうございます！拝見しました！

Commit:
フォロー、アンフォローした際のリダイレクト先の修正。

Commit:
ユーザーをフォローした際のvalidationを修正

Commit:
@userインスタンスを一箇所に定義。relationshipsコントローラーのバリデーションが不適切だったため、修正。

Commit:
フォロー、アンフォローメソッドを利用前にインスタンスを確認しなければならないという意識的な縛りを無くした。

Commit:
ローカル変数としてパーシャルファイルにインスタンス変数を渡すように修正

Commit:
外部キー制約を使用するために、外部キーのカラムをreferences型に変更した。

Commit:
アウトラインとリンクのコードの修正

Commit:
rubocopとlintを通した。

betachelsea commented on app/controllers/follow_relationships_controller.rb:
```diff
@@ -0,0 +1,27 @@
+class FollowRelationshipsController < ApplicationController
+  # フォローするとき
+  def create
+    if User.exists?(id: params[:user_id])
+      @other_user = User.find(params[:user_id])
```
[nits] 2回データベースに確認しにいっているので ( `User.exists?` と `User.find`)、もし問い合わせ回数を減らしたい場合は、次のようにする案もあるかなと思いました〜 👀 

```ruby
if @other_user = User.find_by(id: params[:user_id])
  current_user.follow(@other_user)
  redirect_to @other_user
else
  flash[:alert] = 'このユーザーは存在しません。'
  redirect_to users_path
end
```

betachelsea commented on app/models/follow_relationship.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class FollowRelationship < ApplicationRecord
+  validates :followed, uniqueness: { scope: %i[follower_id followed_id] }
```
[nits] scopeの `followed_id` は実質 `followed` そのものを指すので、特に害は無さそうですが、無効な表記かなと思いました。

```ruby
validates :followed, uniqueness: { scope: :follower_id }
```

で同様の動きをすると思います〜 👀

betachelsea reviewed (approved):
修正ありがとうございました！確認OKです！ 🙆 
少しだけ追加コメントしました〜 📝

Commit:
データベースへの接続が２回行われていたため、これを修正。

バリデーションのスコープも変更。

Commit:
rubocopを通した。
